rustls-mbedpki-provider/src/server_cert_verifier.rs

```diff
@@ -135,35 +135,33 @@ impl ServerCertVerifier for MbedTlsServerCertVerifier {
             .into_iter()
             .collect();
 
-        verify_certificates_active(chain.iter().map(|c| &**c), now, &self.cert_active_check)?;
-
         let server_name_str = server_name_to_str(server_name);
-        let mut error_msg = String::default();
-        match &self.verify_callback {
-            Some(callback) => {
-                let callback = Arc::clone(callback);
-                mbedtls::x509::Certificate::verify_with_callback_expected_common_name(
-                    &chain,
-                    &self.trusted_cas,
-                    None,
-                    Some(&mut error_msg),
-                    move |cert: &mbedtls::x509::Certificate, depth: i32, flags: &mut mbedtls::x509::VerifyError| {
-                        callback(cert, depth, flags)
-                    },
-                    server_name_str.as_deref(),
-                )
-                .map_err(|e| mbedtls_err_into_rustls_err_with_error_msg(e, &error_msg))?;
+
+        let self_verify_callback = self.verify_callback.clone();
+        let callback = move |cert: &mbedtls::x509::Certificate, depth: i32, flags: &mut mbedtls::x509::VerifyError| {
+            // When the "time" feature is enabled for mbedtls, it checks cert expiration. We undo that here,
+            // since this check is done in `verify_certificates_active()` (subject to self.cert_active_check)
+            flags.remove(mbedtls::x509::VerifyError::CERT_EXPIRED | mbedtls::x509::VerifyError::CERT_FUTURE);
+            if let Some(cb) = self_verify_callback.as_ref() {
+                cb(cert, depth, flags)
+            } else {
+                Ok(())
             }
-            None => mbedtls::x509::Certificate::verify_with_expected_common_name(
-                &chain,
-                &self.trusted_cas,
-                None,
-                Some(&mut error_msg),
-                server_name_str.as_deref(),
-            )
-            .map_err(|e| mbedtls_err_into_rustls_err_with_error_msg(e, &error_msg))?,
         };
 
+        let mut error_msg = String::default();
+        mbedtls::x509::Certificate::verify_with_callback_expected_common_name(
+            &chain,
+            &self.trusted_cas,
+            None,
+            Some(&mut error_msg),
+            callback,
+            server_name_str.as_deref(),
+        )
+        .map_err(|e| mbedtls_err_into_rustls_err_with_error_msg(e, &error_msg))?;
+
+        verify_certificates_active(chain.iter().map(|c| &**c), now, &self.cert_active_check)?;
```

Comment:
Similarly, is this intentional to change to verify certificate validity after checking certificate?

Reply:
Again, yes!

Reply:
got it